src/Microsoft.DotNet.Wpf/src/PresentationFramework/Microsoft/Win32/OpenFolderDialog.cs

```diff
@@ -0,0 +1,326 @@
+                _fileNames = ProcessFolders(dialog);
+
+                var cancelArgs = new CancelEventArgs();
+                OnFileOk(cancelArgs);
+                ok = !cancelArgs.Cancel;
+            }
+            finally
+            {
+                if (!ok)
+                {
+                    _fileNames = saveFileNames;
+                }
+            }
+            return ok;
+        }
+
+        #endregion Internal Methods
+
+        //---------------------------------------------------
+        //
+        // Internal Properties
+        //
+        //---------------------------------------------------
+        //#region Internal Properties
+        //#endregion Internal Properties
+
+        //---------------------------------------------------
+        //
+        // Internal Events
+        //
+        //---------------------------------------------------
+        //#region Internal Events
+        //#endregion Internal Events
+
+        //---------------------------------------------------
+        //
+        // Private Methods
+        //
+        //---------------------------------------------------
+        #region Private Methods
+
+        //  Provides the actual implementation of initialization tasks.
+        //  Initialize() is called from both the constructor and the
+        //  public Reset() function to set default values for member
+        //  variables and for the options bitmask.
+        //
+        //  We only perform OpenFolderDialog() specific reset tasks here;
+        //  it's the calling code's responsibility to ensure that the
+        //  base is initialized first.
+        //
+        private void Initialize()
+        {
+            // FOS_FILEMUSTEXIST
+            // Folder must exist. Otherwise, the folder name remaining in
+            // text box might be returned as a nested folder. This flag is
+            // now enforced by FOS_PICKFOLDERS.
+            SetOption(FOS.FILEMUSTEXIST, true);
+
+            // FOS_PICKFOLDERS
+            // This is a folder selecting dialog.
+            SetOption(FOS.PICKFOLDERS, true);
+        }
+
+        #endregion Private Methods
+
+        //---------------------------------------------------
+        //
+        // Private Properties
+        //
+        //---------------------------------------------------
+        #region Private Properties
+
+        /// <summary>
+        ///  Gets a string containing the full path of the file selected in 
+        ///  the file dialog box.
+        /// </summary>
+        private string CriticalFileName
+        {
+            get
+            {
+                if (_fileNames == null)     // No filename stored internally...
+                {
+                    return String.Empty;    // So we return String.Empty
+                }
+                else
+                {
+                    // Return the first filename in the array if it is non-empty.
+                    if (_fileNames[0].Length > 0)
```

Comment:
Do you means `_fileNames.Length > 0`?

```suggestion
                    if (_fileNames.Length > 0)
```

Reply:
List is null when it does not contain entries as comments in line ~140 are telling. It's likely impossible for it to be an empty list at this point. Null check is done above. So, I guess the length check is for the filename and not for the array. But I dont really know in which case it can have an empty string as @miloush said that path & exist check is in place. To my understanding an empty string cannot be a valid path nor file/folder can exist. In which case the check might be redundant.

Reply:
I haven't changed this code, it is the same in FileDialog. I agree this looks wrong, but also in order to trigger the bug you would have to convince the dialog to OK without any files/folders selected, which should not be possible with the current API surface. That said, allowing Multiselect for folders quite complicates things and I might need to review this code.

Reply:
rewritten in the latest commit, thanks @lindexi!